src/server.rs

```diff
@@ -101,9 +99,269 @@ pub enum Event {
+                try!(encoding.write_to(writer));
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Builder of `FramebufferUpdate` message.
+pub struct FramebufferUpdate<'a> {
+    updates: Vec<Update<'a>>,
+}
+
+impl<'a> FramebufferUpdate<'a> {
+    /// Constructs new `FramebufferUpdate`.
+    pub fn new() -> Self {
+        FramebufferUpdate {
+            updates: Vec::new(),
+        }
+    }
+
+    /// Adds raw pixel data.
+    pub fn add_raw_pixels(&mut self, rect: Rect, pixel_data: &'a [u8]) -> &mut Self {
+        let update = Update::Raw {
+            rect: rect,
+            pixel_data: pixel_data
+        };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Adds `CopyRect` update message instructing client to reuse pixel data it already owns.
+    pub fn add_copy_rect(&mut self,
+                         dst: Rect,
+                         src_x_position: u16,
+                         src_y_position: u16)
+                         -> &mut Self {
+        let update = Update::CopyRect {
+            dst: dst,
+            src_x_position: src_x_position,
+            src_y_position: src_y_position,
+        };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Adds compressed pixel data.
+    ///
+    /// TODO: add method taking uncompressed data and compressing them.
+    pub fn add_compressed_pixels(&mut self, rect: Rect, zlib_data: &'a [u8]) -> &mut Self {
+        let update = Update::Zrle {
+            rect: rect,
+            zlib_data: zlib_data
+        };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Add data for drawing cursor.
+    pub fn add_cursor(&mut self,
+                      width: u16,
+                      height: u16,
+                      hotspot_x: u16,
+                      hotspot_y: u16,
+                      pixels: &'a [u8],
+                      mask_bits: &'a [u8])
+                      -> &mut Self {
+        let update = Update::SetCursor {
+            size: (width, height),
+            hotspot: (hotspot_x, hotspot_y),
+            pixels: pixels,
+            mask_bits: mask_bits
+        };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Adds notification about framebuffer resize.
+    pub fn add_desktop_size(&mut self, width: u16, height: u16) -> &mut Self {
+        let update = Update::DesktopSize {
+            width: width,
+            height: height,
+        };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Adds confirmation of support of pseudo-encoding.
+    pub fn add_pseudo_encoding(&mut self, encoding: protocol::Encoding) -> &mut Self {
+        let update = Update::Encoding { encoding: encoding };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Checks if all updates are valid.
+    ///
+    /// Panics if any of the updates is not valid.
+    fn check(&self, validation_data: &ValidationData) {
```

Comment:
Actually I can't borrow anything from server

```
{
    let mut update = server.create_update();
                     ------ immutable borrow occurs here
    update.add_raw_pixels(rect, data);
    server.send_update(&update);
    ^^^^^^ mutable borrow occurs here
}
- immutable borrow ends here
```

To avoid exposing `Update` I would have to add another intermediate struct `FrambufferUpdateBuilder` and use it like this:

```
{
    let update = {
        let mut builder = server.create_update();
        builder.add_raw_pixels(rect, data);
        builder.done()
    };
    server.send_update(&update);
}
```